### DISCUSSION TITLE
Two simple encryption/archive plugins I'm happy to share...

### DISCUSSION THREAD

**Discussion by ArchieLinux** [jarun/nnn#1567] [Show and tell]
I wrote two simple bash scripts for my own use and posted them publicly to share:

https://github.com/ArchieLinux/nnn-plugs

In short, "`arcenc`" and "`arcdec`" are a pair of simple scripts to (1) compress/encrypt and (2) decompress/decrypt, respectively,  a file or directory for safe private archiving, emailing as an attachment, etc. 

They both use `tar` and `gpg` in series, providing excellent compression & robust encryption. As such, these are the two dependencies of which I'm aware. Leveraging the `gpg` "silence" feature, passwords are **not** echo'ed to avoid potential leakage of encryption keys.

Being a big fan of `nnn`, I wanted to be able to accomplish this through the `nnn` file manager - and in turn thought others might enjoy them, too. Of course, they can also be run as external CLI bash scripts. As far as I know, they are POSIX compliant - on my best efforts basis. 

**I would be honored to have them included in the nnn library of plugins under The Open Unlicense.**

**Comment by jarun** [comment]
👍 

@N-R-K and @luukvbaal could you please evaluate?

**Comment by KlzXS** [comment]
I like the idea. Here are some observations I have:

- The tar command uses gzip and can't be changed to another algorithm. Also it always uses the tar format.
- You only support symmetric encryption. What about asymmetric?
- There already exist plugins (gpge, gpgd) for gpg encryption and decryption (why's there no plugin for signing and signature verification 🤦 ?) which support both symmetric and asymmetric encryption.
- They support key selection using fzf which is a dependency your scripts don't have.
- They also don't do any preprocessing to the passed files, just encrypt and decrypt. They're not capable of the workflow you describe on their own.

**Comment by ArchieLinux** [comment]
@KlzXS Many thanks for your prompt review and evaluation. On each point...

- I hadn't really seen the need to offer a range of compression formats besides how `tar` functions - would that useful for others' use cases?
- Very good point; since I don't use private/public keys, extra menu choices around such features would slow down my "getting to the finish line" given my own typical use case. Would this feature be in high demand by others?
- Wow! TBH, being new to `nnn` your existing plugins totally escaped my notice. My apologies! Having read your scripts, they seem to be extremely well-constructed, such that I might not have even bothered to write my own had I spotted them. Kudos :+1: 
- I'm not sure how involving `fzf` would be much of a benefit as one is typically well aware of which target files to archive/encrypt and where they reside (?)
- Yes, combining both archiving + encryption together was one of my key objectives. While compression is not always essential based on what's being archived, a directory chock full of uncompressed files - which case is an archive/encryption situation I often face - can make a meaningful difference.

Which brings me to a (very quick) trial run of `gpge` and `gpgd`. I will study both more closely to learn more about your  thoughtful CLI scripting. 

FWIW, I didn't seem to be able to apply `gpge` to a directory. Perhaps I need to look more closely at what I'm doing, but for me being able to archive/encrypt an entire directory is an important ingredient. No doubt it's just user error on my part, so I will try to apply `gpge` to a directories and see how I can do this.

Thanks again for your insightful comments. I will definitely spend more time on this in the near future. :vulcan_salute:

**Reply by KlzXS** [reply]
To address you concerns:

- I personally prefer xz over gzip. And even 7z sometimes. Why? I have no idea, a nicer name I'd say. When in doubt, I always ere on the side of broad coverage.
- The previous point can apply to using keys for encryption. I personally don't encrypt stuff that often, and when I do it's usually indirectly through a third part program. In theory the keys provide a stronger encyption, more bits/entropy and the such. High demand? Don't know, not part of that sphere.
- `fzf` is there to allow the selection of the private key to use. One might have multiple private keys for multiple uses. There are alternative ways to do that, but this is the prettiest.
- If the plugins don't work on directories that would be a limitation of `gpg` itself. And that sounds about right.

**Reply by ArchieLinux** [reply]
Thanks again for the additional comments/insights. 

FWIW, the `arcenc`/`arcdec` plugins do indeed function on directories using `gpg` as they recognize and preserve the full tree structure recursively. Being fairly certain `gpg` can do this perhaps it's worth a closer look. That particular feature is important to how I tend to create and use archives. :shrug:

**Comment by amalgame21** [comment]
Nice to see someone try to implement archive plugin for public use.
I have wrote some plugin for that with interactive options.
Those are just for my own use, bug may exist, many duplicated logic and complex, may not portable to other system like MacOS.
You may test it out and adopt it to your own script.

https://github.com/amalgame21/conifg/blob/main/archive_tar
https://github.com/amalgame21/conifg/blob/main/archive_extractor

`archive_tar`, interactive archive using bsdtar and zstd/lz4 compression (you may add gzip / xz support), with optional symmetric gpg (no asymmetric key since I don't use it often, you can add it to the script.). Use `bat` for better display of what is going on, you may change to `less`.
`archive_extractor`, extract archive like nnn does, but support selection (not tested for same directory name/file name inside different archives, may delete existing files, use with your own risk).

To decrypt and extract `tar.zst.gpg` created, i just use `gpgd` , then with built-in nnn extraction (with atool with zstd support)

**Reply by ArchieLinux** [reply]
Thanks for the helpful encouragement!

FYI, I just added another plugin for `nnn` that I'm happy to share. Would be honored to have it included here under the jarun's repository. Otherwise you can find it here: https://github.com/ArchieLinux/nnn-plugs

I'll add a fresh "discussion" item here since it serves an entirely different purpose,
